### ISSUE TITLE
Symfony v6 Controller getDoctrine()

### ISSUE THREAD

**Issue by 94noni** [rectorphp/rector-symfony#94]
Hi

According to https://github.com/symfony/symfony/blob/6.0/UPGRADE-6.0.md, `getDoctrine()` is being deprecated
I have a lot of code `$em = $this->getDoctrine()->getManager();` in my controllers, I would like to know if there is, or how to add, a rule to swap from this code to `EntityManagerInterface $entityManager` constructor injection or controller action injection

thank you

**Comment by TomasVotruba** [comment]
Hi, could you share before and after example?

**Comment by 94noni** [comment]
Hi thanks for reply
Yes I ll post back soon :)

**Comment by 94noni** [comment]
@TomasVotruba here you are:

```diff

// src/Controller/SomeController.php

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+ use Doctrine\ORM\EntityManagerInterface;

 class SomeController extends AbstractController 
 {
-    public function index()
+    public function index(EntityManagerInterface $em)
     {
-         $em = $this->getDoctrine()->getManager();
           $em->getRepository(...);
     }
 }
```

`$em` injected because the var was named `$em` in the controller code

**Comment by TomasVotruba** [comment]
The constructor injection would be better, but I get the idea. Thanks :+1: 

That's all methods or is there any other diff you can give me?

**Comment by 94noni** [comment]
I've written example of controller action injection as this project is an evolution of an old symfony v2 app to v3,4,5 now 6
so old controller with lot of actions methods

On some new controllers, I prefer one invoke action with constructor injection that is why I showed an exemple of action injection, as I know rector can already do constructor injection

for what I can read, there is also the `dispatchMessage()` of the same class from symfony :)

**Comment by TomasVotruba** [comment]
> so old controller with lot of actions methods

It might be handy to split such god class with custom simple Rector rule :+1: 

> On some new controllers, I prefer one invoke action with constructor injection that is why I showed an exemple of action injection, as I know rector can already do constructor injection

The thing is, we have internal services for adding dependencies to constructor. We can re-use those easily. For any other modification, you can add custom rule to move your dependencies to places of need.

> for what I can read, there is also the dispatchMessage() of the same class from symfony :)

Could you share diff for that one?

**Comment by 94noni** [comment]
@TomasVotruba, for the latest part, kind of similar to the entity manager interface but for the messenger bus:

```diff

// src/Controller/SomeController.php

use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+ use Symfony\Component\Messenger\MessageBusInterface;

 class SomeController extends AbstractController 
 {
-    public function index()
+    public function index(MessageBusInterface $bus)
     {
-         $bus = $this->dispatchMessage(...);
           $bus->dispatch(...);
     }
 }
```

`$bus` injected because the var was named `$bus` in the controller's action code 

---

yes I need to improve this "old" code behavior split with more new logic, but I am new to rector I am testing step by step ;)

**Comment by TomasVotruba** [comment]
> But I am new to rector I am testing step by step ;) and so far so good, this tool is a must have thank you so much

That's the best approach. Start slowly then build up every day step by step :) thank you for positive feedback :+1: 

---

I'm looking into it... related pull-requests:

* https://github.com/symfony/symfony/pull/42442
* https://github.com/symfony/symfony/pull/1195

**Comment by TomasVotruba** [comment]
Fixed in #97 ad #99

**Comment by 94noni** [comment]
I will be pleased to test it live and report back, will check the next release :)
thank you @TomasVotruba
